### PR TITLE
Restore pre-`3.28` behavior for `which_package`

### PR DESCRIPTION
**Pull request by mbargull**
### Description

Fix linking check regressions by restoring pre-`3.28` behavior for `which_package`.

Fixes gh-5136.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] ~~Add / update outdated documentation?~~

### CODE REVIEW

File: conda_build/inspect_pkg.py [hunk 1]
```diff
@@ -67,43 +67,21 @@ def which_package(
...
-    except FileNotFoundError:
-        # FileNotFoundError: path doesn't exist
-        return
-    else:
-        yield from _file_package_mapping(prefix).get(lstat, ())
-
+        path = Path(path).relative_to(prefix)
+    except ValueError:
+        # If path is already relative to prefix, we get a ValueError.
+        pass
```
Comment (mbargull) [inline conda_build/inspect_pkg.py:75]: I don't know why this was changed to take paths not relative to the prefix in 3.28 .
To keep the change set small, I just added handling of non-relatives paths here instead of at `which_packages` call sites.

File: conda_build/inspect_pkg.py [hunk 2]
```diff
@@ -67,43 +67,21 @@ def which_package(
...
+    # On Windows, be lenient and allow case-insensitive path comparisons.
+    # NOTE: On macOS, although case-insensitive filesystem is default, still
+    #       require case-sensitive matches (i.e., normcase on macOS is a no-op).
+    normcase_path = normcase(str(path).replace(os.sep, "/"))
```
Comment (mbargull) [inline conda_build/inspect_pkg.py:79]: In the current 3.28 versions, there was a behavioral change in that builds on a case-insensitive filesystem on Unix systems (i.e., macOS with default file system) would ignore case mismatches.
Pre-3.28 behavior was to check for case-sensitivity on all Unixes (since `normcase`'s implementation only ignores cases on Windows).

File: tests/conftest.py [hunk 3]
```diff
@@ -227,7 +227,13 @@ def conda_build_test_recipe_path(tmp_path_factory: pytest.TempPathFactory) -> Pa
     # clone conda_build_test_recipe locally
     repo = tmp_path_factory.mktemp("conda_build_test_recipe", numbered=False)
     subprocess.run(
-        ["git", "clone", "https://github.com/conda/conda_build_test_recipe", str(repo)],
+        [
+            "git",
+            "clone",
+            "https://github.com/conda/conda_build_test_recipe",
+            "--branch=1.21.11",
```
Comment (mbargull) [inline tests/conftest.py:234]: This is an unrelated change to make the tests pass.
It should probably not be picked verbatim for the `main` branch.

File: tests/test_inspect_pkg.py [hunk 4]
```diff
@@ -101,7 +129,11 @@ def test_which_package(tmp_path: Path):
     precs_missing = list(which_package(tmp_path / "missing", tmp_path))
     assert not precs_missing
 
-    precs_hardlinkA = list(which_package(tmp_path / "hardlinkA", tmp_path))
+    if on_win:
+        # On Windows, be lenient and allow case-insensitive path comparisons.
+        precs_hardlinkA = list(which_package(tmp_path / "Hardlinka", tmp_path))
+    else:
+        precs_hardlinkA = list(which_package(tmp_path / "hardlinkA", tmp_path))
```
Comment (mbargull) [inline tests/test_inspect_pkg.py:136]: See https://github.com/conda/conda-build/pull/5141/files#r1454195421 above:
`conda-build<3.28` would only allow case-insensitive matches on Windows.

File: tests/test_inspect_pkg.py [hunk 5]
```diff
@@ -172,10 +212,14 @@ def test_which_package_battery(tmp_path: Path):
...
-    # this occurs when, e.g., a package removes files installed by another package
+    # removed files should still return a package
+    # this occurs when, e.g., a build script removes files installed by another package
+    # (The opposite case with files really missing from the run environment,
+    # e.g., due to a post-install script removing them, is less likely and not
+    # covered.)
+    #  covered here.)
     for file in removed:
-        assert not len(list(which_package(tmp_path / file, tmp_path)))
+        assert len(list(which_package(tmp_path / file, tmp_path))) == 1
```
Comment (mbargull) [inline tests/test_inspect_pkg.py:222]: This change is a direct consequence of https://github.com/conda/conda-build/pull/5141/files#r1454205167 above, i.e., files may be removed during build from the `host` prefix.

**Comment by mbargull** [comment]
@kenodegard, I reviewed the code for `which_package` for <=3.27` and found some more behavioral changes that `3.28.x` introduced.
See the code as well as review comments in this PR for details.

I think we should revert the function's behavior to (mostly) what was done in `<3.27` and put out a new `3.28` patch release for it.
We should discuss further changes (behavioral and/or performance-wise) for future versions rather than the `3.28.x` line.

File: conda_build/inspect_pkg.py (hunk 1 above)
Comment (kenodegard) [reply conda_build/inspect_pkg.py:75]: yeah this change came about because I saw how before every call to `which_package` we would strip the prefix, so inlining the relative logic here makes sense

File: conda_build/inspect_pkg.py (hunk 2 above)
Comment (kenodegard) [reply conda_build/inspect_pkg.py:79]: from Python docs on `os.path.normcase`:
> On Windows, convert all characters in the pathname to lowercase, and also convert forward slashes to backward slashes.

```suggestion
    # On Windows, be lenient and allow case-insensitive path comparisons.
    # NOTE: On macOS, although case-insensitive filesystem is default, still
    #       require case-sensitive matches (i.e., normcase on macOS is a no-op).
    normcase_path = normcase(path)
```

File: tests/test_inspect_pkg.py [hunk 6]
```diff
@@ -24,15 +25,27 @@ def test_which_package(tmp_path: Path):
     (tmp_path / "internal").symlink_to(tmp_path / "hardlinkA")  # packageA
     (tmp_path / "external").symlink_to(tmp_path / "hardlinkB")  # packageA
     (tmp_path / "hardlinkB").touch()  # packageB
+    # Files might be deleted from the prefix during the build, but they should
+    # still be recognized since they will be present in the run environment.
+    (tmp_path / "deleted").unlink(missing_ok=True)  # packageA
+    (tmp_path / "deleted_shared").unlink(missing_ok=True)  # packageA & packageB
```
[the remark this answers is not in the document]
Comment (kenodegard) [reply tests/test_inspect_pkg.py:31]: ah I see, a mistake on my part in understanding where overlinking detection happened

File: tests/conftest.py (hunk 3 above)
Comment (kenodegard) [reply tests/conftest.py:234]: since we're about to start the January releases I'm thinking we will want to change the target branch and revert this change

Comment (mbargull) [reply tests/conftest.py:234]: Do you mean this commit in particular or the whole PR?
I'm totally fine to drop the commit -- it was just a means to get "all green".
For the PR itself, I'd like us to have another `3.28` patch release if possible.
The thing is that we don't have any working `3.28` package on conda-forge (some are already marked `broken`) and I'd like us to get one such release before we introduce further changes from the upcoming version.
Meaning, as a maintainer of infrastructure-related packages on conda-forge, I'd like to do one of the following:
1. Put out a build for a new `3.28` patch release once tagged upstream.
2. Add this PR as a patch to the recipe for the current version to get a working downstream build.
3. Mark all `3.28` builds as `broken` and thus continue to use `3.27` for now on conda-forge.

Option 1 being the most preferred.
Option 2 is not ideal since we'd have divergent behavior up- and downstream.
Option 3 would be the last resort to get things working again.
(For conda-forge we'd rather want to increase not decrease the use of linking checks and omitting the linking checks via config would only create maintenance burden and is thus not an option I'm willing to go with.)

File: tests/test_inspect_pkg.py [hunk 7]
```diff
@@ -172,10 +212,14 @@ def test_which_package_battery(tmp_path: Path):
...
             assert len(list(which_package(path, tmp_path))) == 1
 
-    # removed files should return no packages
-    # this occurs when, e.g., a package removes files installed by another package
+    # removed files should still return a package
+    # this occurs when, e.g., a build script removes files installed by another package
+    # (The opposite case with files really missing from the run environment,
+    # e.g., due to a post-install script removing them, is less likely and not
+    # covered.)
+    #  covered here.)
```
Comment (kenodegard) [inline tests/test_inspect_pkg.py:220]:
```suggestion
    # (post-install scripts removing files from the run environment is less
    # likely and not covered)
```

File: news/5141-fix-which_package [hunk 8]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Fix linking check regressions by restoring pre-3.28 behavior for which_package. (#5141)
```
Comment (kenodegard) [inline news/5141-fix-which_package:7]:
```suggestion
* Fix linking check regressions by restoring pre-3.28 behavior for `conda_build.inspect_pkg.which_package`. (#5141)
```

File: conda_build/inspect_pkg.py [hunk 9]
```diff
@@ -67,43 +67,21 @@ def which_package(
...
-            try:
-                lstat = (prefix / file).lstat()
-            except FileNotFoundError:
-                # FileNotFoundError: path doesn't exist
-                continue
-            else:
-                mapping.setdefault(lstat, set()).add(prec)
-    return mapping
+        files = prec["files"]
+        if normcase_path in (normcase(file) for file in files):
```
Comment (kenodegard) [inline conda_build/inspect_pkg.py:83]:
```suggestion
        if normcase_path in (normcase(file) for file in prec["files"]):
```

File: conda_build/inspect_pkg.py [hunk 10]
```diff
@@ -67,43 +67,21 @@ def which_package(
...
-        lstat = (prefix / path).lstat()
-    except FileNotFoundError:
-        # FileNotFoundError: path doesn't exist
-        return
-    else:
-        yield from _file_package_mapping(prefix).get(lstat, ())
-
+        path = Path(path).relative_to(prefix)
+    except ValueError:
+        # If path is already relative to prefix, we get a ValueError.
```
Comment (kenodegard) [inline conda_build/inspect_pkg.py:74]:
```suggestion
        # ValueError: path is already relative to prefix
```

**Review by kenodegard — APPROVED** [review]
minor suggestions

File: tests/conftest.py (hunk 3 above)
Comment (kenodegard) [reply tests/conftest.py:234]: working with the team to figure out a release plan

added to 24.1.x milestone only because this is something that will need to be resolved and merged before the release can proceed

**Review by kenodegard — APPROVED** [review]
I added the regression test I wrote in #5140, otherwise LGTM

**Comment by minrk** [comment]
Thanks for the fix!
